src/List/ListItemText.js

```diff
@@ -41,54 +41,60 @@ function ListItemText(props, context) {
     className: classNameProp,
     disableTypography,
     inset,
-    primary = children,
-    secondary,
+    primary: primaryProp,
+    secondary: secondaryProp,
     ...other
   } = props;
   const { dense } = context;
-  const className = classNames(
-    classes.root,
-    {
-      [classes.dense]: dense,
-      [classes.inset]: inset,
-    },
-    classNameProp,
-  );
+
+  let primary = primaryProp || children;
+  if (primary && !disableTypography) {
+    primary = (
+      <Typography
+        variant="subheading"
+        className={classNames(classes.primary, { [classes.textDense]: dense })}
+      >
+        {primary}
+      </Typography>
+    );
+  }
+
+  let secondary = secondaryProp;
+  if (secondary && !disableTypography) {
+    secondary = (
+      <Typography
+        variant="body1"
+        className={classNames(classes.secondary, {
+          [classes.textDense]: dense,
+        })}
+        color="textSecondary"
+      >
+        {secondary}
+      </Typography>
+    );
+  }
 
   return (
-    <div className={className} {...other}>
-      {primary &&
-        (disableTypography ? (
-          primary
-        ) : (
-          <Typography
-            variant="subheading"
-            className={classNames(classes.primary, { [classes.textDense]: dense })}
-          >
-            {primary}
-          </Typography>
-        ))}
-      {secondary &&
-        (disableTypography ? (
-          secondary
-        ) : (
-          <Typography
-            variant="body1"
-            className={classNames(classes.secondary, {
-              [classes.textDense]: dense,
-            })}
-            color="textSecondary"
-          >
-            {secondary}
-          </Typography>
-        ))}
+    <div
+      className={classNames(
+        classes.root,
+        {
+          [classes.dense]: dense,
+          [classes.inset]: inset,
+        },
+        classNameProp,
+      )}
+      {...other}
+    >
+      {primary}
+      {secondary}
```

Comment:
hehe I wanted to refactor that too, but more like 

```js
const textChildren = disableTypography ? (
  <>
    {primary}
    {secondary}
  </>
) : (
  <>
    {primary && (
      <Typography
        variant="subheading"
        className={classNames(classes.primary, { [classes.textDense]: dense })}
      >
        {primary}
      </Typography>
    )}
    {secondary && (
      <Typography
        variant="body1"
        className={classNames(classes.secondary, {
          [classes.textDense]: dense,
        })}
        color="textSecondary"
      >
        {secondary}
      </Typography>
    )}
  </>
);
```
(or `<React.Fragment>` for `<>` not sure it would work without babel7+)

Reply:
We can't use the fragment yet. We are still supporting React 15.x.